### PR TITLE
Minor updates for TRestGDMLParser

### PR DESCRIPTION
**Pull request by lobis**
lobis Medium: 147 [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/lobis-activeVolume-43)

* Related to https://github.com/rest-for-physics/restG4/issues/43
* Replaces Nevents by nEvents in RML files (Nevents still works) for consistency
* Replaces gdml_file by gdmlFile in RML files (gdml_file still works) for consistency

TRestGDMLParser:
* Renamed members for consistency (starting with `f`...)
* Made some methods parameters by const ref.
* When USER is not defined, will save temporary gdml file to `/tmp` directory instead of failing
* Includes <filesystem> header to create directory recursively if not exists
* Other minor changes

### CODE REVIEW

File: cmake/thisREST.cmake [hunk 1]
```diff
@@ -15,7 +15,7 @@ string(REGEX REPLACE "\n$" "" GEANT4_PATH "${GEANT4_PATH}")
 set(thisGeant4 "${GEANT4_PATH}/bin/geant4.sh")
 
 if (${REST_G4} MATCHES "ON")
-    set(loadG4 "\# if geant4.sh script is found we load the same Geant4 version as used in compilation\n if [[ -f \"${thisGeant4}\" && ${thisGeant4} != /usr/* ]]; then
+    set(loadG4 "\# if geant4.sh script is found we load the same Geant4 version as used in compilation\n if [[ -f \\\"${thisGeant4}\\\" && ${thisGeant4} != /usr/* ]]; then
```
Comment (jgalan) [inline cmake/thisREST.cmake:18]: is this a bug fix?

File: source/framework/core/inc/TRestGDMLParser.h [hunk 2]
```diff
@@ -7,49 +7,51 @@
...
 
     void PrintContent();
 
     void ReplaceEntity();
 
-    void ReplaceAttributeWithKeyWord(std::string keyword);
+    void ReplaceAttributeWithKeyWord(const std::string& keyword);
 
-    ClassDef(TRestGDMLParser, 1);
+    ClassDefOverride(TRestGDMLParser, 2);
```
Comment (jgalan) [inline source/framework/core/inc/TRestGDMLParser.h:54]: For curiosity, when should we use `DefOverride` and why?

File: source/framework/core/src/TRestMetadata.cxx [hunk 3]
```diff
@@ -515,8 +515,8 @@ TRestMetadata::TRestMetadata(const char* cfgFileName) : endl(this) {
 /// \brief TRestMetadata default destructor
 ///
 TRestMetadata::~TRestMetadata() {
-    if (fElementGlobal) delete fElementGlobal;
-    if (fElement) delete fElement;
+    delete fElementGlobal;
```
Comment (jgalan) [inline source/framework/core/src/TRestMetadata.cxx:518]: why not this check? It is not bad to check if the pointer is `nullptr` then just do nothing in that case.

**Review by jgalan — APPROVED** [review]

**Review by juanangp — APPROVED** [review]

File: cmake/thisREST.cmake (hunk 1 above)
Comment (lobis) [reply cmake/thisREST.cmake:18]: Not really sure, it was introduced by @juanangp

File: source/framework/core/src/TRestMetadata.cxx (hunk 3 above)
Comment (lobis) [reply source/framework/core/src/TRestMetadata.cxx:518]: > why not this check? It is not bad to check if the pointer is `nullptr` then just do nothing in that case.

Deleting a null pointer is perfectly safe, so there is not need to check this.

https://stackoverflow.com/questions/615355/is-there-any-reason-to-check-for-a-null-pointer-before-deleting

File: cmake/thisREST.cmake (hunk 1 above)
Comment (juanangp) [reply cmake/thisREST.cmake:18]: Yes, this is a bug fix, it was giving an error after performing `cmake`

File: source/framework/core/src/TRestMetadata.cxx (hunk 3 above)
Comment (jgalan) [reply source/framework/core/src/TRestMetadata.cxx:518]: The main problem is that delete will not remove the actual pointer address. This is something I discussed with @juanangp already in 2015. So, it is necessary to set it to null to make sure you wont access it again or you don't believe it is existing. Probably at the constructor it does nothing.

**Pull request merged**